File: 백준/Silver/9012. 괄호/괄호.py
T = int(input())
for tc in range(1,T+1):
    s = input()
    top = 0
    for i in s:
        if i == '(':
            top += 1
        elif i == ')':
             top -=1
        if top < 0:
            break
    if top == 0:
        print('YES')
    else:
        print('NO')

File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int main(void)
{
	int N;
	cin >> N;
	for (int k = 0; k < N; k++)
	{
		stack <char> s;
		string inp;
		cin >> inp;
		for (int i = 0; i < inp.length(); i++)
		{
			if (inp[i] == '(')
			{
				s.push(inp[i]);
			}
			else
			{
				if (s.empty() || s.top() != '(')
				{
					s.push(inp[i]);
					break;
				}
				else
				{
					s.pop();
				}
			}
		}
		if (s.empty())
		{
			cout << "YES" << '\n';
		}
		else
		{
			cout << "NO" << '\n';
		}
	}
	
	return 0;
}